File: app/src/main/java/giuliolodi/gitnav/GistListActivity.java
/*
 * Copyright 2016 GLodi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package giuliolodi.gitnav;


import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;

public class GistListActivity extends BaseDrawerActivity {

    @BindView(R.id.tab_layout) TabLayout tabLayout;
    @BindView(R.id.gists_viewpager) ViewPager gistsViewpager;

    @BindString(R.string.mine) String mine;
    @BindString(R.string.starred) String starred;
    @BindString(R.string.network_error) String network_error;

    private GistListMine gistListMine;
    private GistListStarred gistListStarred;
    private List<Integer> views;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getLayoutInflater().inflate(R.layout.gist_list_activity, frameLayout);

        ButterKnife.bind(this);

        views = new ArrayList<>();
        views.add(R.layout.gist_list_mine);
        views.add(R.layout.gist_list_starred);

        gistsViewpager.setOffscreenPageLimit(2);
        gistsViewpager.setAdapter(new MyAdapter(getApplicationContext()));

        tabLayout.setVisibility(View.VISIBLE);
        tabLayout.setSelectedTabIndicatorColor(Color.WHITE);
        tabLayout.setupWithViewPager(gistsViewpager);

    }

    public class MyAdapter extends PagerAdapter {

        Context context;

        public MyAdapter(Context context) {
            this.context = context;
        }

        @Override
        public int getCount() {
            return 2;
        }

        @Override
        public boolean isViewFromObject(View view, Object object) {
            return view.equals(object);
        }

        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            ViewGroup layout = (ViewGroup) getLayoutInflater().inflate(views.get(position), container, false);
            container.addView(layout);
            return layout;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case 0:
                    return mine;
                case 1:
                    return starred;
            }
            return super.getPageTitle(position);
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
            container.removeView((View) object);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        navigationView.getMenu().getItem(4).setChecked(true);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        if (Constants.isNetworkAvailable(getApplicationContext())) {
            gistListMine = new GistListMine();
            gistListMine.populate(GistListActivity.this, findViewById(R.id.gists_mine_rl));
            gistListStarred = new GistListStarred();
            gistListStarred.populate(GistListActivity.this, findViewById(R.id.gists_starred_rl));
        } else
            Toast.makeText(getApplicationContext(), network_error, Toast.LENGTH_LONG).show();
        return true;
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        overridePendingTransition(0,0);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_options:
                startActivity(new Intent(getApplicationContext(), OptionActivity.class));
                overridePendingTransition(0,0);
            default:
                return super.onOptionsItemSelected(item);
        }
    }

}

File: app/src/main/java/giuliolodi/gitnav/ui/gists/GistListActivity.kt
/*
 * Copyright 2017 GLodi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package giuliolodi.gitnav.ui.gists

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.support.v4.view.PagerAdapter
import android.view.*
import android.widget.Toast
import giuliolodi.gitnav.ui.base.BaseDrawerActivity
import kotlinx.android.synthetic.main.activity_base_drawer.*
import org.eclipse.egit.github.core.Gist
import javax.inject.Inject
import giuliolodi.gitnav.R.string.network_error
import es.dmoral.toasty.Toasty
import giuliolodi.gitnav.R
import giuliolodi.gitnav.utils.NetworkUtils
import kotlinx.android.synthetic.main.app_bar_home.*
import kotlinx.android.synthetic.main.gist_list_activity.*

/**
 * Created by giulio on 23/05/2017.
 */

class GistListActivity : BaseDrawerActivity(), GistListContract.View {

    @Inject lateinit var mPresenter: GistListContract.Presenter<GistListContract.View>

    private val mViews: MutableList<Int> = arrayListOf()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        layoutInflater.inflate(R.layout.gist_list_activity, content_frame)

        mViews.add(R.layout.gist_list_mine)
        mViews.add(R.layout.gist_list_starred)

        gist_list_viewpager.offscreenPageLimit = 2
        gist_list_viewpager.adapter = GistListAdapter(applicationContext, mViews)

        tab_layout.visibility = View.VISIBLE
        tab_layout.setSelectedTabIndicatorColor(Color.WHITE)
        tab_layout.setupWithViewPager(gist_list_viewpager)
    }

    override fun showMineGists(gistList: List<Gist>) {
    }

    override fun showStarredGists(gistList: List<Gist>) {
    }

    override fun showLoadingMine() {
    }

    override fun showLoadingStarred() {
    }

    override fun hideLoadingMine() {
    }

    override fun hideLoadingStarred() {
    }

    override fun showError(error: String) {
    }

    private  class GistListAdapter(context: Context, views: List<Int>) : PagerAdapter() {

        private var mContext = context
        private var mViews = views

        override fun instantiateItem(container: ViewGroup?, position: Int): Any {
            val layout = LayoutInflater.from(mContext).inflate(mViews[position], container, false)
            container?.addView(layout)
            return layout
        }

        override fun getPageTitle(position: Int): CharSequence {
            when (position) {
                0 -> return mContext.getString(R.string.mine)
                1 -> return mContext.getString(R.string.starred)
            }
            return super.getPageTitle(position)
        }

        override fun destroyItem(container: ViewGroup?, position: Int, `object`: Any?) {
            container?.removeView(`object` as View)
        }

        override fun isViewFromObject(view: View?, `object`: Any?): Boolean {
            return view?.equals(`object`)!!
        }

        override fun getCount(): Int {
            return 2
        }

    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        if (NetworkUtils.isNetworkAvailable(applicationContext)) {

        } else
            Toasty.warning(applicationContext, getString(network_error), Toast.LENGTH_LONG).show()
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        nav_view.menu.getItem(5).isChecked = true
    }

    override fun onDestroy() {
        mPresenter.onDetach()
        super.onDestroy()
    }

    override fun onBackPressed() {
        super.onBackPressed()
        overridePendingTransition(0,0)
    }

    companion object {
        fun getIntent(context: Context): Intent {
            return Intent(context, GistListActivity::class.java)
        }
    }
}